cdap-credential-ext-gcp-wi/src/main/java/io/cdap/cdap/security/spi/credential/GcpWorkloadIdentityCredentialProvider.java

```diff
@@ -0,0 +1,326 @@
+
+    this.client = Config.defaultClient();
+    int connectTimeoutSec = Integer.parseInt(
+        credentialProviderContext.getProperties().get(CONNECT_TIMEOUT_SECS));
+    int readTimeoutSec = Integer.parseInt(
+        credentialProviderContext.getProperties().get(READ_TIMEOUT_SECS));
+    OkHttpClient httpClient = this.client.getHttpClient().newBuilder()
+        .connectTimeout(connectTimeoutSec, TimeUnit.SECONDS)
+        .readTimeout(readTimeoutSec, TimeUnit.SECONDS)
+        .build();
+    this.client.setHttpClient(httpClient);
+
+    return this.client;
+  }
+
+  @Override
+  public ProvisionedCredential provision(NamespaceMeta namespaceMeta,
+      CredentialProfile profile, CredentialIdentity identity)
+      throws CredentialProvisioningException {
+
+    // Provision the credential with exponential delay on retryable failure.
+    long delay = Long.parseLong(
+        credentialProviderContext.getProperties().get(RETRY_BASE_DELAY_MILLIS));
+    long maxDelay = Long.parseLong(
+        credentialProviderContext.getProperties().get(RETRY_MAX_DELAY_MILLIS));
+    long timeout = Long.parseLong(
+        credentialProviderContext.getProperties().get(RETRY_TIMEOUT_MILLIS));
+    double minMultiplier = RETRY_DELAY_MULTIPLIER - RETRY_DELAY_MULTIPLIER * RETRY_RANDOMIZE_FACTOR;
+    double maxMultiplier = RETRY_DELAY_MULTIPLIER + RETRY_DELAY_MULTIPLIER * RETRY_RANDOMIZE_FACTOR;
+    Stopwatch stopWatch = Stopwatch.createStarted();
+    try {
+      while (stopWatch.elapsed(TimeUnit.SECONDS) < timeout) {
+        try {
+          return getProvisionedCredential(namespaceMeta, identity);
+        } catch (RetryableException e) {
+          TimeUnit.MILLISECONDS.sleep(delay);
+          delay = (long) (delay * (minMultiplier + Math.random() * (maxMultiplier - minMultiplier
+              + 1)));
+          delay = Math.min(delay, maxDelay);
+        } catch (Exception e) {
+
+          LOG.error(
+              String.format(PROVISIONING_FAILURE_ERROR_MESSAGE_FORMAT, identity.getIdentity()), e);
+
+          throw new CredentialProvisioningException(
+              String.format(PROVISIONING_FAILURE_ERROR_MESSAGE_FORMAT, identity.getIdentity()), e);
+        }
+      }
+    } catch (InterruptedException e) {
+      throw new CredentialProvisioningException(
+          String.format(PROVISIONING_FAILURE_ERROR_MESSAGE_FORMAT, identity.getIdentity()), e);
+    }
+
+    // timed out while provisioning the credential.
+    throw new CredentialProvisioningException(
+        new TimeoutException(
+            String.format("Timed out while provisioning the credential for identity '%s'",
+                identity.getIdentity())
+        ));
+  }
+
+  private ProvisionedCredential getProvisionedCredential(NamespaceMeta namespaceMeta,
+      CredentialIdentity identity) throws IOException, ApiException {
+
+    // get k8s namespace from namespace metadata.
+    String k8sNamespace = namespaceMeta.getConfig().getConfigs().getOrDefault("k8s.namespace",
```

Comment:
Should we also check that namespace creation is enabled? Otherwise (even though unlikely), this property set on a namespace by a user on an instance without namespace creation enabled will cause issues. As a suggestion, we could do this by adding it to the context as a boolean.

Reply:
done.